CMake build system - WIP

slokhorst opened the pull request:
This is a follow-up to #229, to show what the CMake build system would approximately look like. Do not merge!

Build with:

``` bash
cmake .
make
```

Some remarks/questions:
- This requires #230
- Installation (`make install`) does not work yet.
- What is the difference between `PACKAGE_VERSION` and `VERSION`? Do we need two distinct variables?
- I have removed all autotools stuff, to show the difference between the two systems. However, we can easily put that back and provide both options.

jogu commented:
Cool, thanks! I'll give this a go at some point. From looking through it, it looks a lot more readable and a lot less verbose than configure.ac, which hopefully makes it less error prone.

If I'm reading correctly, pkg-config isn't used for finding ffmpeg in the new way? I think I would prefer if we continued to do so if possible. Using pkg-config hopefully means it'll both get picked up if it's been installed in a weird place, and means we don't have to track as ffmpeg add/remove sublibraries as they refactor.

slokhorst commented:
Ah, yes, pkg-config is much better! Completely forgot about that. Replaces the entire `FindFFMPEG.cmake` file with one simple line. :)

jogu commented:
Oh, btw, to answer your question "What is the difference between PACKAGE_VERSION and VERSION? Do we need two distinct variables?". I'm not sure - I don't see any difference so to my eye it looks like they can be combined.

jogu commented:
I get an error when I run cmake without mysql available (on MacOS, but I don't think that part matters):

-- Could NOT find MMAL (missing:  MMAL_LIBRARIES MMAL_INCLUDE_DIRS)
-- Could NOT find MYSQL (missing:  MYSQL_LIBRARY MYSQL_INCLUDE_DIR)
-- Could NOT find PostgreSQL (missing:  PostgreSQL_INCLUDE_DIR PostgreSQL_TYPE_INCLUDE_DIR)
-- Could NOT find SDL (missing:  SDL_LIBRARY SDL_INCLUDE_DIR)
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
MYSQL_INCLUDE_DIR (ADVANCED)
   used as include directory in directory /Users/joseph/Documents/motion
   used as include directory in directory /Users/joseph/Documents/motion
   used as include directory in directory /Users/joseph/Documents/motion
MYSQL_LIBRARY (ADVANCED)
    linked by target "motion" in directory /Users/joseph/Documents/motion

This is a /lot/ faster than configure btw!

slokhorst commented:
@jogu, sorry about that. Didn't test properly without the optional components. I have added options to enable/disable the optional components, but I'm not sure what to use as default options. I think we have two options (or a combination of both):
* First search for all components, and enable everything that is available
* Disable everything and require manual enabling of features

The first might be easier for manual building, so the user has the most features possible, but the second is likely preferable for distro/package maintainers, as it is more predictable.

I'm finding it hard to determine what the current behavior with `configure` is.

BTW: You can use the options like so: `cmake -DWITH_MYSQL=OFF .`, or use the interactive `ccmake .` tool.

jogu commented:
Great, thanks! I'll hopefully be able to try that out tonight.

The approach with the current configure is (as far as I'm aware) to enable anything that the user has all the dependencies for.

The official Debian package is currently configured with:
          --with-ffmpeg=$(LIBDIR_MULTIARCH) 
	  --with-mysql-lib=$(LIBDIR_MULTIARCH) 
	  --with-pgsql-lib=$(LIBDIR_MULTIARCH)
	  --without-optimizecpu
(see https://sources.debian.net/src/motion/3.4.1-1/debian/rules/#L12 )

I think this effectively forces the system to error out if the ffmpeg/mysql/pgsql dependencies aren't available. I guess it would be ideal if we could have equivalent options that could be passed to cmake.

slokhorst commented:
Ok, it currently is set to "build everything that we can build", but everything can be manually turned on or off, like so (similar to e.g. [Blender](https://sources.debian.net/src/blender/2.77.a%2Bdfsg0-9/debian/rules/)):

````bash
cmake . \
    -DWITH_FFMPEG=[ON|OFF] \
    -DWITH_MMAL=[ON|OFF] \
    -DWITH_MYSQL=[ON|OFF] \
    -DWITH_PGSQL=[ON|OFF] \
    -DWITH_PTHREAD=[ON|OFF] \
    -DWITH_SDL=[ON|OFF] \
    -DWITH_SQLITE3=[ON|OFF]
````

jogu commented:
Thanks @slokhorst ! I can build okay on MacOS with the latest changes (the binary doesn't work very well, but that's a separate issue :-) ).

I've not yet discovered how I do a build with symbols included (-g) and optimisations disabled (-O0) with cmake (so that I can usefully use gdb/lldb) - do you know?

slokhorst commented:
@jogu You can do that with `CMAKE_C_FLAGS`. You can set it on the command line:
```bash
cmake . -DCMAKE_C_FLAGS='-O0 -g'
````

The flags set in `CMakeLists.txt` (line 5) will be appended to the ones you set like this.

jogu commented on CMakeLists.txt:
```diff
@@ -0,0 +1,121 @@
+
+execute_process(COMMAND sh "${CMAKE_CURRENT_SOURCE_DIR}/version.sh" OUTPUT_VARIABLE VERSION)
+set(BIN_PATH "${CMAKE_INSTALL_PREFIX}/bin")
+set(sysconfdir "/etc" CACHE STRING "system configuration directory")
```
I think this differs from the behaviour of the current setup which defaults to /usr/local/etc - we should probably keep the default as /usr/local/etc ?

slokhorst replied:
Good point, fixed.

jogu commented:
One other issue I ran into that I'm not sure if it's fixable - when I use cmake to generate an Xcode project on MacOS, the path to the ffmpeg libraries isn't added to the linker path so it fails to link.

I think that might be a cmake bug: https://gitlab.kitware.com/cmake/cmake/issues/16154 - not sure if we can workaround it.

Oddly it seems to work okay when generating a makefile, not sure why.

Mr-Dave commented:
Sorry I've been diverted on other things.  Couple of comments:    
1.  First, this is really fast.  :)  
2.  There was a question above regarding defaults.  If the option is installed on the system, we should have it enabled.  If it isn't installed, we disable the option and continue.  The exception to this is ffmpeg.  The configuration should fail if ffmpeg is not installed and the user did not disable it with the configuration option.  i.e. It is required unless specifically disabled.    
3. The mmal testing appears to be "borrowed" from kodi.  I'm not sure what the protocol is on this to make sure that they get proper developer accreditation.  Somehow we need to remove or revise the line that says this file is part of xbmc while still making sure that the developer gets proper credit.  
4.  I think that there may be a bug in the install.  When changing the install prefix, it isn't revising the path for the config files.  All the paths to the other files seem to be changed correctly.  
5.  There are unusual configuration options in the old configuration regarding sqlite3 where it sets <code>HAVE_SQLITE3_EMBEDDED</code>.  Not sure about the feasibility of replicating the historic behaviour.

jogu commented:
> 5. There are unusual configuration options in the old configuration regarding sqlite3 where it sets HAVE_SQLITE3_EMBEDDED. Not sure about the feasibility of replicating the historic behaviour.

I think I'd be quite keen to drop this support. It's not documented, and I can't really see any good reason to use this rather than the OS provided sqlite libraries, suggesting it may be unnecessary complexity. We can re-consider if someone comes forward that can explain why it's worth having.

One other thing with cmake: It seems that I don't get any symbols in coredumps since switching to cmake. I guess it may be stripping the binaries by default. I'm not necessarily sure either way is better, but just noting that the default behaviour has changed.

slokhorst commented:
> The configuration should fail if ffmpeg is not installed and the user did not disable it with the configuration option. 

Done: a2cf2ec

> The mmal testing appears to be "borrowed" from kodi.

Yes, that's true, I got it from [here](https://github.com/xbmc/xbmc/blob/master/project/cmake/modules/FindMMAL.cmake). I actually added those lines myself to show where it was from. We are free to include it, as Kodi is licensed under GPL2, but I'm not sure what the usual way to credit the author is. 

> HAVE_SQLITE3_EMBEDDED

I think you should be able to work around this with pkg-config, and point to a nonstandard installation directory. Also: it's a bit strange to have this embedded option just for sqlite, and not for the other libraries.

> I think that there may be a bug in the install. When changing the install prefix, it isn't revising the path for the config files.

I also saw this. I'm looking for a solution.

> One other thing with cmake: It seems that I don't get any symbols in coredumps since switching to cmake. I guess it may be stripping the binaries by default. I'm not necessarily sure either way is better, but just noting that the default behaviour has changed.

I'm not sure about this. I'll try to look into it.

jogu commented:
My thoughts are that (subject to working through the things people have already mentioned) this is now pretty close to being mergeable.

I think once we've ironed out those last few things we just need to put back the all the autoconf files and we can merge this, start encouraging people to use it to bring to light any remaining differences.
(It might be worth doing a quick git rebase -i to tidy up some of the commits - e.g. removing the commit that got reverted later.)

Thanks for all your work on this @slokhorst, very much appreciated!

Mr-Dave commented:
I concur with joseph, this is great work!.

slokhorst commented:
Sorry for the lack of action, I was quite busy.

I have rebased this against current master, put back the automake stuff, and squashed everything into one commit.

* I think the current accreditation for the file copied from Kodi is ok. We clearly name the author, the licence used, and where we got it. Strangely enough, I can't find a lot of "etiquette" online for how to do this properly, but I think this is good enough. If you think of anything better, please change it/tell me to change it before merging.
* The problem with `sysconfdir` is that the following happens: On the first run of `cmake`, it defaults to `${CMAKE_INSTALL_PREFIX}/etc`, but if `CMAKE_INSTALL_PREFIX` is changed in a subsequent run of `cmake -DCMAKE_INSTALL_PREFIX=...`, the value of `sysconfdir` is not updated as it's cached. We cache it because it is also possible to set `sysconfdir` with `cmake -Dsysconfdir=...`, and it's preferable that you don't need to do that on every run.. I think?
* I'm not sure what to do about the Travis file. Should we build the test with automake or CMake? Or test them both? Currently I've reverted it to automake.

If you agree with the copyright notice in the Kodi file, I propose we merge this. The `sysconfdir` issue is easy to work around, and if we think of a proper fix, we can always apply it later.

jogu commented:
Thanks @slokhorst !

I think this looks good to merge.

Importing the GPL2 file from kodi seems fine; we've got the original copyright and a note of the license and where it came from.

For the travis file: I suspect we should have at least one cmake build defined in it, so that no one accidentally breaks cmake builds. I hope we can fully switch to cmake within a few months if no problems arise.

jogu commented:
I added a wiki page here to try and cover a quick intro to how to build with cmake;

https://github.com/Motion-Project/motion/wiki/cmake

This'll need to be migrated into the guide, INSTALL, etc in due course.

tosiara commented:
I wonder why cmake requires C++ compiler?
```
CMake Error at CMakeLists.txt:1 (project):
  No CMAKE_CXX_COMPILER could be found.
```

jogu commented:
It looks like cmake doesn't pick up v4l on my system. Seems the current configure.ac was setting -DMOTION_V4L2 and -DMOTION_V4L2_OLD on my machine. I'm unsure what "OLD" means in this context as I'm running kernel 3.16 which is pretty recent. I don't actually use any video4linux cameras so never noticed this before.

jogu commented:
@tosiara I think if you change the first line to this it should fix that:

` project(motion C)`

tosiara commented:
Yes, that fixed. Should it be there by default?
Another question - how do I specify `ffmpeg` location for cmake?

slokhorst commented:
@tosiara regarding ffmpeg:
We use pkg-config to locate ffmpeg.

If you installed ffmpeg like `./configure --prefix=$ffprefix && make && make install` then the ffmpeg pkg-config files are located in `$ffprefix/`

To have CMake/pkg-config find them, use the following:
````
PKG_CONFIG_PATH=$ffprefix/lib/pkgconfig cmake .
````

slokhorst commented:
@jogu regarding `MOTION_V4L2` and `MOTION_V4L2_OLD`, they are not being set at all, among a [few other variables](https://github.com/Motion-Project/motion/blob/master/config.h.in#L18) (I didn't need them (and couldn't test them) to get my V4L2 webcam working on Arch Linux). Which distro are you using? Then I'll try to reimplement those options to get it working there.

jogu commented:
@slokhorst I'm using Debian jessie x86_64.

jogu commented:
@tosiara Yes, it would seem reasonable to have that by default - please feel free to submit a pull request :)

tosiara commented:
Now getting this:
```
PKG_CONFIG_PATH=/home/user/git/ffmpeg/out/lib/pkgconfig cmake .
CMake Error: Could not find cmake module file: CMakeDeterminecCompiler.cmake
CMake Error: Error required internal CMake variable not set, cmake may be not be built correctly.
Missing variable is:
CMAKE_c_COMPILER_ENV_VAR
CMake Error: Error required internal CMake variable not set, cmake may be not be built correctly.
Missing variable is:
CMAKE_c_COMPILER
CMake Error: Could not find cmake module file: /home/user/git/motion/CMakeFiles/3.3.2/CMakecCompiler.cmake
CMake Error at CMakeLists.txt:1 (project):
  No CMAKE_c_COMPILER could be found.

  Tell CMake where to find the compiler by setting the CMake cache entry
  CMAKE_c_COMPILER to the full path to the compiler, or to the compiler name
  if it is in the PATH.


CMake Error: Could not find cmake module file: CMakecInformation.cmake
CMake Error: CMAKE_c_COMPILER not set, after EnableLanguage
-- Configuring incomplete, errors occurred!
See also "/home/user/git/motion/CMakeFiles/CMakeOutput.log".
```

slokhorst commented:
@tosiara  I think you have added `c` (lower case) instead of `C` to the `project()` line in `CMakeLists.txt`

tosiara commented:
True, my fault :) Used lowercase. Now getting this error during make:
```
# make
Scanning dependencies of target motion
[  4%] Building C object CMakeFiles/motion.dir/conf.c.o
[  8%] Building C object CMakeFiles/motion.dir/motion.c.o
In file included from /home/user/git/motion/motion.c:9:0:
/home/user/git/motion/ffmpeg.h:13:34: fatal error: libavformat/avformat.h: No such file or directory
 #include <libavformat/avformat.h>
                                  ^
compilation terminated.
CMakeFiles/motion.dir/build.make:86: recipe for target 'CMakeFiles/motion.dir/motion.c.o' failed
make[2]: *** [CMakeFiles/motion.dir/motion.c.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/motion.dir/all' failed
make[1]: *** [CMakeFiles/motion.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

slokhorst commented:
@tosiara You need ffmpeg or similar. For Debian that's the following packages: `libav-tools libavcodec-dev libavformat-dev libswscale-dev`. Alternatively you can run `cmake -DWITH_FFMPEG=OFF .` to disable FFmpeg.

tosiara commented:
I do have ffmpeg built from sources. And I'm trying to compile motion against that as I always do with 'old' configure way. I have specified path to ffmpeg during cmake:
# PKG_CONFIG_PATH=/home/user/git/ffmpeg/out/lib/pkgconfig cmake .

Mr-DaveDev commented:
This has actually already been merged so we need to discuss the further enhancements in the issue.  Note that I think there is more work to do with the manually specified paths.

slokhorst commented:
@tosiara sorry, forgot about that. As MrDave said, lets continue this in #229.

slokhorst commented:
(sorry for continuing here, but this is a comment specifically about the discussion above)

@tosiara can you post the contents of `/home/user/git/ffmpeg/out/lib/pkgconfig/libavformat.pc`? Especially the first `prefix=` line?

tosiara commented:
Damn, I have just realized I have done the same mistake twice: https://github.com/Motion-Project/motion/issues/143
I again used relative path for ffmpeg prefix and this was causing the issue:
```
# cat ~/git/ffmpeg/out/lib/pkgconfig/libavformat.pc
prefix=out
exec_prefix=${prefix}
libdir=${prefix}/lib
includedir=${prefix}/include

Name: libavformat
Description: FFmpeg container format library
Version: 57.46.100
Requires: libavcodec >= 57.51.100, libswresample >= 2.1.100, libavutil >= 55.28.100
Requires.private: 
Conflicts:
Libs: -L${libdir}  -lavformat -lx265 -lm -pthread
Libs.private: 
Cflags: -I${includedir}
```
So now I have rebuilt ffmpeg with a correct prefix, applied your patch and moved a bit forward. This time I get:
```
[ 91%] Building C object CMakeFiles/motion.dir/vloopback_motion.c.o
[ 95%] Building C object CMakeFiles/motion.dir/webhttpd.c.o
[100%] Linking C executable motion
/usr/lib64/gcc/x86_64-suse-linux/4.8/../../../../x86_64-suse-linux/bin/ld: cannot find -lavformat
/usr/lib64/gcc/x86_64-suse-linux/4.8/../../../../x86_64-suse-linux/bin/ld: cannot find -lavcodec
/usr/lib64/gcc/x86_64-suse-linux/4.8/../../../../x86_64-suse-linux/bin/ld: cannot find -lswresample
/usr/lib64/gcc/x86_64-suse-linux/4.8/../../../../x86_64-suse-linux/bin/ld: cannot find -lswscale
/usr/lib64/gcc/x86_64-suse-linux/4.8/../../../../x86_64-suse-linux/bin/ld: cannot find -lavutil
/usr/lib64/gcc/x86_64-suse-linux/4.8/../../../../x86_64-suse-linux/bin/ld: cannot find -lavcodec
/usr/lib64/gcc/x86_64-suse-linux/4.8/../../../../x86_64-suse-linux/bin/ld: cannot find -lswresample
/usr/lib64/gcc/x86_64-suse-linux/4.8/../../../../x86_64-suse-linux/bin/ld: cannot find -lswscale
/usr/lib64/gcc/x86_64-suse-linux/4.8/../../../../x86_64-suse-linux/bin/ld: cannot find -lavutil
collect2: error: ld returned 1 exit status
CMakeFiles/motion.dir/build.make:667: recipe for target 'motion' failed
make[2]: *** [motion] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/motion.dir/all' failed
make[1]: *** [CMakeFiles/motion.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

slokhorst commented:
Hm, I was also having the linking issue, but 3a0f007714b10383987ff8fc99278c0a730d55b7 fixed that.
This is just a wild guess, but maybe running `rm -r CMakeCache.txt CMakeFiles` (inside of the motion directory, so it won't have any of the old paths in the cache), and then trying again, will help?

tosiara commented:
Took a fresh machine, cloned clean repo - still the same. Full log:
```
odroid@odroidc1:~/git/motion$ PKG_CONFIG_PATH=/home/odroid/git/ffmpeg/out/lib/pkgconfig cmake .
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found JPEG: /usr/lib/arm-linux-gnueabihf/libjpeg.so
-- Could NOT find MMAL (missing:  MMAL_LIBRARIES MMAL_INCLUDE_DIRS)
-- Found MYSQL: /usr/lib/arm-linux-gnueabihf/libmysqlclient.so
-- Could NOT find PostgreSQL (missing:  PostgreSQL_LIBRARY PostgreSQL_INCLUDE_DIR PostgreSQL_TYPE_INCLUDE_DIR)
-- Could NOT find SDL (missing:  SDL_LIBRARY SDL_INCLUDE_DIR)
-- Checking for module 'sqlite3'
--   No package 'sqlite3' found
-- Looking for include file linux/videodev.h
-- Looking for include file linux/videodev.h - not found
-- Looking for include file linux/videodev2.h
-- Looking for include file linux/videodev2.h - found
-- Looking for include file sys/videoio.h
-- Looking for include file sys/videoio.h - not found
-- Checking for modules 'libavutil;libavformat;libavcodec;libswscale'
--   Found libavutil, version 55.40.100
--   Found libavformat, version 57.57.100
--   Found libavcodec, version 57.66.103
--   Found libswscale, version 4.3.101
-- Configuring done
-- Generating done
-- Build files have been written to: /home/odroid/git/motion
odroid@odroidc1:~/git/motion$ make
Scanning dependencies of target motion
[  4%] Building C object CMakeFiles/motion.dir/conf.c.o
[  8%] Building C object CMakeFiles/motion.dir/motion.c.o
[ 12%] Building C object CMakeFiles/motion.dir/alg.c.o
[ 16%] Building C object CMakeFiles/motion.dir/draw.c.o
[ 20%] Building C object CMakeFiles/motion.dir/event.c.o
[ 25%] Building C object CMakeFiles/motion.dir/ffmpeg.c.o
/home/odroid/git/motion/ffmpeg.c: In function Б─≤ffmpeg_openБ─≥:
/home/odroid/git/motion/ffmpeg.c:412:5: warning: Б─≤codecБ─≥ is deprecated [-Wdeprecated-declarations]
     ffmpeg->c     = c = AVSTREAM_CODEC_PTR(ffmpeg->video_st);
     ^
In file included from /home/odroid/git/motion/ffmpeg.h:13:0,
                 from /home/odroid/git/motion/ffmpeg.c:29:
/home/odroid/git/ffmpeg/out/include/libavformat/avformat.h:893:21: note: declared here
     AVCodecContext *codec;
                     ^
/home/odroid/git/motion/ffmpeg.c: In function Б─≤ffmpeg_cleanupsБ─≥:
/home/odroid/git/motion/ffmpeg.c:571:9: warning: Б─≤codecБ─≥ is deprecated [-Wdeprecated-declarations]
         avcodec_close(AVSTREAM_CODEC_PTR(ffmpeg->video_st));
         ^
In file included from /home/odroid/git/motion/ffmpeg.h:13:0,
                 from /home/odroid/git/motion/ffmpeg.c:29:
/home/odroid/git/ffmpeg/out/include/libavformat/avformat.h:893:21: note: declared here
     AVCodecContext *codec;
                     ^
/home/odroid/git/motion/ffmpeg.c: In function Б─≤ffmpeg_closeБ─≥:
/home/odroid/git/motion/ffmpeg.c:592:9: warning: Б─≤codecБ─≥ is deprecated [-Wdeprecated-declarations]
         avcodec_close(AVSTREAM_CODEC_PTR(ffmpeg->video_st));
         ^
In file included from /home/odroid/git/motion/ffmpeg.h:13:0,
                 from /home/odroid/git/motion/ffmpeg.c:29:
/home/odroid/git/ffmpeg/out/include/libavformat/avformat.h:893:21: note: declared here
     AVCodecContext *codec;
                     ^
/home/odroid/git/motion/ffmpeg.c: In function Б─≤ffmpeg_put_frameБ─≥:
/home/odroid/git/motion/ffmpeg.c:716:9: warning: Б─≤avcodec_encode_video2Б─≥ is deprecated [-Wdeprecated-declarations]
         retcd = avcodec_encode_video2(AVSTREAM_CODEC_PTR(ffmpeg->video_st),
         ^
In file included from /home/odroid/git/ffmpeg/out/include/libavformat/avformat.h:319:0,
                 from /home/odroid/git/motion/ffmpeg.h:13,
                 from /home/odroid/git/motion/ffmpeg.c:29:
/home/odroid/git/ffmpeg/out/include/libavcodec/avcodec.h:5322:5: note: declared here
 int avcodec_encode_video2(AVCodecContext *avctx, AVPacket *avpkt,
     ^
/home/odroid/git/motion/ffmpeg.c:716:9: warning: Б─≤codecБ─≥ is deprecated [-Wdeprecated-declarations]
         retcd = avcodec_encode_video2(AVSTREAM_CODEC_PTR(ffmpeg->video_st),
         ^
In file included from /home/odroid/git/motion/ffmpeg.h:13:0,
                 from /home/odroid/git/motion/ffmpeg.c:29:
/home/odroid/git/ffmpeg/out/include/libavformat/avformat.h:893:21: note: declared here
     AVCodecContext *codec;
                     ^
[ 29%] Building C object CMakeFiles/motion.dir/jpegutils.c.o
[ 33%] Building C object CMakeFiles/motion.dir/logger.c.o
[ 37%] Building C object CMakeFiles/motion.dir/md5.c.o
[ 41%] Building C object CMakeFiles/motion.dir/netcam.c.o
[ 45%] Building C object CMakeFiles/motion.dir/netcam_ftp.c.o
[ 50%] Building C object CMakeFiles/motion.dir/netcam_jpeg.c.o
[ 54%] Building C object CMakeFiles/motion.dir/netcam_rtsp.c.o
/home/odroid/git/motion/netcam_rtsp.c: In function Б─≤decode_packetБ─≥:
/home/odroid/git/motion/netcam_rtsp.c:102:5: warning: Б─≤avcodec_decode_video2Б─≥ is deprecated [-Wdeprecated-declarations]
     retcd = avcodec_decode_video2(cc, frame, &check, packet);
     ^
In file included from /home/odroid/git/motion/netcam_rtsp.h:8:0,
                 from /home/odroid/git/motion/netcam_rtsp.c:23:
/home/odroid/git/ffmpeg/out/include/libavcodec/avcodec.h:4811:5: note: declared here
 int avcodec_decode_video2(AVCodecContext *avctx, AVFrame *picture,
     ^
/home/odroid/git/motion/netcam_rtsp.c: In function Б─≤netcam_open_codecБ─≥:
/home/odroid/git/motion/netcam_rtsp.c:160:5: warning: Б─≤codecБ─≥ is deprecated [-Wdeprecated-declarations]
     dec_ctx = st->codec;
     ^
In file included from /home/odroid/git/motion/netcam_rtsp.h:9:0,
                 from /home/odroid/git/motion/netcam_rtsp.c:23:
/home/odroid/git/ffmpeg/out/include/libavformat/avformat.h:893:21: note: declared here
     AVCodecContext *codec;
                     ^
/home/odroid/git/motion/netcam_rtsp.c: In function Б─≤netcam_rtsp_open_contextБ─≥:
/home/odroid/git/motion/netcam_rtsp.c:488:5: warning: Б─≤codecБ─≥ is deprecated [-Wdeprecated-declarations]
     netcam->rtsp->codec_context = netcam->rtsp->format_context->streams[netcam->rtsp->video_stream_index]->codec;
     ^
In file included from /home/odroid/git/motion/netcam_rtsp.h:9:0,
                 from /home/odroid/git/motion/netcam_rtsp.c:23:
/home/odroid/git/ffmpeg/out/include/libavformat/avformat.h:893:21: note: declared here
     AVCodecContext *codec;
                     ^
[ 58%] Building C object CMakeFiles/motion.dir/netcam_wget.c.o
[ 62%] Building C object CMakeFiles/motion.dir/picture.c.o
[ 66%] Building C object CMakeFiles/motion.dir/rotate.c.o
[ 70%] Building C object CMakeFiles/motion.dir/stream.c.o
[ 75%] Building C object CMakeFiles/motion.dir/track.c.o
[ 79%] Building C object CMakeFiles/motion.dir/video2.c.o
[ 83%] Building C object CMakeFiles/motion.dir/video.c.o
[ 87%] Building C object CMakeFiles/motion.dir/video_common.c.o
/home/odroid/git/motion/video_common.c: In function Б─≤vid_nextБ─≥:
/home/odroid/git/motion/video_common.c:979:20: warning: variable Б─≤heightБ─≥ set but not used [-Wunused-but-set-variable]
         int width, height;
                    ^
/home/odroid/git/motion/video_common.c:979:13: warning: variable Б─≤widthБ─≥ set but not used [-Wunused-but-set-variable]
         int width, height;
             ^
[ 91%] Building C object CMakeFiles/motion.dir/vloopback_motion.c.o
[ 95%] Building C object CMakeFiles/motion.dir/webhttpd.c.o
[100%] Linking C executable motion
/usr/bin/ld: cannot find -lavformat
/usr/bin/ld: cannot find -lavcodec
/usr/bin/ld: cannot find -lswresample
/usr/bin/ld: cannot find -lswscale
/usr/bin/ld: cannot find -lavutil
/usr/bin/ld: cannot find -lavcodec
/usr/bin/ld: cannot find -lswresample
/usr/bin/ld: cannot find -lswscale
/usr/bin/ld: cannot find -lavutil
collect2: error: ld returned 1 exit status
CMakeFiles/motion.dir/build.make:669: recipe for target 'motion' failed
make[2]: *** [motion] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/motion.dir/all' failed
make[1]: *** [CMakeFiles/motion.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

jogu commented:
@tosiara This might possibly be https://gitlab.kitware.com/cmake/cmake/issues/16154#note_188104 :(

tosiara commented:
In other words, CMake system currently does not allow to build motion using custom ffmpeg libs?

tosiara commented:
Let's move further discussion to a dedicated thread: https://github.com/Motion-Project/motion/issues/257
